src/transport/raw/TCP.cpp

```diff
@@ -290,101 +268,90 @@ CHIP_ERROR TCPBase::SendAfterConnect(const PeerAddress & addr, System::PacketBuf
+    size_t consumed = 0;
 
     while (!buffer.IsNull())
     {
-        // when a buffer is empty, it can be released back to the app
-        if (buffer->DataLength() == 0)
-        {
-            buffer.FreeHead();
-            continue;
-        }
-
-        uint8_t * messageData = nullptr;
-        uint16_t messageSize  = 0;
-        if (ContainsCompleteMessage(buffer, &messageData, &messageSize))
+        uint8_t messageSizeBuf[kPacketSizeBytes];
+        if (buffer->Read(messageSizeBuf) != CHIP_NO_ERROR)
         {
-            // length was read and is not needed anymore
-            buffer->ConsumeHead(kPacketSizeBytes);
-
-            // Sanity checks. These are more like an assert for invariants
-            VerifyOrExit(messageData == buffer->Start(), err = CHIP_ERROR_INTERNAL);
-            VerifyOrExit(buffer->DataLength() >= messageSize, err = CHIP_ERROR_INTERNAL);
-
-            // messagesize is always consumed once processed, even on error. This is done
-            // on purpose:
-            //   - we already consumed the packet size above
-            //   - there is no reason to believe that an error would not occur again on the
-            //     same parameters (errors are likely not transient)
-            //   - this guarantees data is received and progress is made.
-            err = ProcessSingleMessageFromBufferHead(peerAddress, buffer, messageSize);
-            buffer->ConsumeHead(messageSize);
-            SuccessOrExit(err);
-
-            err = endPoint->AckReceive(messageSize);
-            SuccessOrExit(err);
-            continue;
+            break;
         }
-
-        // Buffer is incomplete if we reach this point
-        if (buffer->HasChainedBuffer())
+        uint16_t messageSize = LittleEndian::Get16(messageSizeBuf);
+        if (messageSize >= (UINT16_MAX - kPacketSizeBytes))
         {
-            buffer->CompactHead();
-            continue;
+            // Trying to accumulate a message this long would make TotalLength() overflow.
+            // TODO: Issue #5438 - Close TCP connection when errors prevent handling messages
+            return CHIP_ERROR_MESSAGE_TOO_LONG;
         }
-
-        if (messageSize > 0)
+        // The subtraction will not underflow because we successfully read kPacketSizeBytes.
+        if (messageSize <= (buffer->TotalLength() - kPacketSizeBytes))
         {
-            // Open the receive window just enough to allow the remainder of the message to be received.
-            // This is necessary in the case where the message size exceeds the TCP window size to ensure
-            // the peer has enough window to send us the entire message.
-            uint16_t neededLen = static_cast<uint16_t>(messageSize - buffer->DataLength());
-            err                = endPoint->AckReceive(neededLen);
-            SuccessOrExit(err);
+            // The buffer contains a complete message.
+            consumed += messageSize + kPacketSizeBytes;
```

Comment:
I think ultimately, consumed has to always be how much came in. Otherwise you may end-up with having insufficient window to receive the entire message is the messageSize is >= TCP receive window size. This means, ultimately, that I think it's necessary to enqueue the incoming buffers in here, to dequeue them from TCPEndpoint, or to know how much data is coming in on the batch, irrespective of the yet-to-be-consumed chain in TCPEndPoint.

Reply:
It looks to me like LWIP will always append a new pbuf in `TCPEndPoint::HandleDataReceived`, and socket platforms can expand the receive queue in `TCPEndPoint::ReceiveData` (managing acks and window with the underlying calls to recv).  The only missing thing then is to always `tcp_recved` in LWIP (`TCPEndPoint::AckReceive`) to make sure that there is at least one byte available from LWIP `TCP_WND`.

The only reason not to ack immediately would be to apply back-pressure at the application layer in the LWIP platform.  However, as you allude, one needs to make sure there is always a way for the window to open back up if it closes to zero.

So for LWIP specifically, I agree with the diagnosis: with the code as written, the window can close to zero and permanently prevent forward progress.  However, we don't necessarily need to consume bytes to solve this.  The TCP endpoint and this code can share the buffer until it contains a complete message.  What we do need to do is ACK in LWIP to keep the window open.  If we were to choose to ever return from this context with the window closed and without acking (this is possible with the current code), for LWIP, we would need to set a timer or similar to reexamine conditions later and ack at that time.  That's more complex of course.

TLDR: in LWIP, it seems like we should always just ack all bytes received, not all bytes consumed.

The refactor that removed the buffer expansion from this context seems like a good one.  The TCP stack already does this for us.  However, I think we lost the ACKing of partial messages and we need that back.

Reply:
Yes, the next commit will ack as soon as we receive the buffer in `ProcessReceivedBuffer()`. (This should be up later today; I'm just beefing up the unit test.)

Reply:
FWIW, the LWIP ack can be accomplished very simply by calling it instead from `TCPEndPoint::HandleDataReceived`:

```
diff --git a/src/inet/TCPEndPoint.cpp b/src/inet/TCPEndPoint.cpp
index 181a8afa1..296b67342 100644
--- a/src/inet/TCPEndPoint.cpp
+++ b/src/inet/TCPEndPoint.cpp
@@ -2007,6 +2007,9 @@ void TCPEndPoint::HandleDataReceived(System::PacketBufferHandle buf)
 
         // Drive the received data into the app.
         DriveReceiving();
+       // #5625 suggestion: How about lets just LWIP ack here.  Unless we actually intend to apply
+       // back-pressure from the consuming code (difficult to do correctly), this will always work.
+       // It is also fucntionally equivalent to socket-platform behavior.
     }
 }
```

Reply:
The reason not to do this is that it allows unbounded queuing if the device is busy. ACK being done after having at least looked at incoming queue to see if messages can be handled allows a critical throttling

Reply:
Summarizing what's added in subsequent commits, I see quite of cleanup; ACK is almost where suggested, but before the potential DoClose ([here](https://github.com/project-chip/connectedhomeip/pull/5265/commits/3e8a7c7c6d3949c746ffc5e8e50963c0472c376c#L1391)):

```
void TCPEndPoint::DriveReceiving()
...
        AckReceive(ackLength);
    }
   if (State == kState_Closing && mRcvQueue.IsNull())
        DoClose(INET_NO_ERROR, false);
}
```

and very importantly, no attacker can enqueue very large messages ([here](https://github.com/project-chip/connectedhomeip/pull/5265/commits/52b76b3d785586741610bb1d21a03a09f4f3cc5a#L315)):

```
        if (messageSize >= System::PacketBuffer::kMaxSizeWithoutReserve)
        {
            // This message is too long for upper layers, which require a contiguous buffer.
            // TODO: Actual limit may be lower (spec issue #2119)
            state->mReceived.Consume(kPacketSizeBytes);
            state->mDiscardSize = messageSize;
            ReturnErrorOnFailure(state->Discard());
            return CHIP_ERROR_MESSAGE_TOO_LONG;
        }
```

Looks good to me.  I won't hit the merge button yet in case anybody has any final input.